---
title: "Arunachalam et al. 2020 dataset"
date: "`r format(Sys.time(), '%Y %B %d')`"
output:
    html_notebook:
        self_contained: true
        df_print: paged
        fig_height: 6
        fig_width: 6
        highlight: zenburn
        theme: cosmo
        number_sections: yes
        toc: yes
        toc_depth: 6
        toc_float:
            collapsed: false
---

```{r setup}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE
)
```

```{r}
library(GEOquery)
library(Matrix)
library(tidyverse)
library(Seurat)
library(harmony)
library(DT)
library(fgsea)
library(patchwork)
library(ComplexHeatmap)
library(RColorBrewer)
library(future)
library(scater)
library(edgeR)
library(gghighlight)
library(ggrepel)
library(data.table)
library(ggpubr)
library(scales)
library(dendsort)

show_dt <- function(df) { 
    DT::datatable(
        df,
        class = "cell-border stripe",
        rownames = TRUE,
        escape = FALSE,
        caption = NULL,
        editable = FALSE,
        filter = list(position = "top"),
        extensions = c('Buttons', 'Scroller'),
        options = list(
            dom = 'Blfrtip',
            scrollX = T,
            scrollY = 500,
            buttons = c('copy', 'csv', 'print'),
            deferRender = TRUE,
            scroller = TRUE
        )
    )
}

theme_set(theme_classic(14))
```

```{r}
gmt.hallmark <- gmtPathways("h.all.v7.2.symbols.gmt")
gmt.btm <- gmtPathways("BTMgenesets.gmt")
```

# Load data and pre-processing

Use the getGEOSuppFiles function to download data from GEO (GSE155673).

```{r}
filePaths <- getGEOSuppFiles("GSE155673")
```

Set as working directory the folder where data have been downloaded and place the file 'samples.csv' in this folder.

```{r}
setwd("GSE155673/") 

mat.list <- dir(pattern = "*.mtx")
mat <- readMM(mat.list[1])

for(i in mat.list[-1]) {
    mat <- Matrix::cBind(mat, readMM(i))
}

mat.genes <- read_tsv("GSE155673_features.tsv.gz", col_names = F)
rownames(mat) <- mat.genes$X2

barcodes.list <- dir(pattern = "GSE155673_cov.*_barcodes.tsv")
barcodes <- lapply(barcodes.list, function(x) read_tsv(x, col_names = F)$X1) 

samples <- read_csv("samples.csv") %>%
    add_column(n_barcodes = sapply(barcodes, length))

barcodes <- barcodes %>% unlist()

colnames(mat) <- paste(barcodes, rep(samples$patient_id, samples$n_barcodes), sep = "_")

seu <- CreateSeuratObject(counts = mat[mat.genes$X3== "Gene Expression", ], project = "pulendran", assay = "RNA", min.cells = 0, min.features = 0, meta.data = NULL)
seu[["ADT"]] <- CreateAssayObject(counts = mat[mat.genes$X3 == "Antibody Capture", ])

seu <- AddMetaData(seu, metadata = samples %>% dplyr::slice(rep(1:n(), times = samples$n_barcodes)) %>% select(-n_barcodes) %>% add_column(id = colnames(seu)) %>% as.data.frame() %>% column_to_rownames("id"))

rm(list = c("barcodes", "barcodes.list", "i", "mat", "mat.genes", "mat.list", "samples"))
```

```{r}
seu <- subset(seu, cells = colnames(seu)[rowSums(GetAssayData(seu, assay = "RNA", slot = "counts")) > 2])
```

```{r}
dim(seu)
```

```{r}
seu <- PercentageFeatureSet(seu, features = grep("^MT\\-", rownames(seu), value = T), col.name = "percent.mt")
```

```{r fig.height=4, fig.width=10}
VlnPlot(seu, assay = "RNA", slot = "counts", features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), same.y.lims = F, log = F, pt.size = 0, ncol = 3)
```

```{r}
seu <- subset(seu, subset = nFeature_RNA >= 500 & percent.mt < 10)
dim(seu)
```

```{r message=FALSE, warning=FALSE}
plan("sequential")
seu <- seu %>%
    SCTransform(assay = "RNA", new.assay.name = "SCT", do.correct.umi = T, vars.to.regress = c("nCount_RNA", "percent.mt"), variable.features.n = 3000, variable.features.rv.th = NULL, return.only.var.genes = T, conserve.memory = T, seed.use = 123) %>%
    RunPCA(assay = "SCT", npcs = 30) %>%
    RunHarmony(group.by.vars = c("patient_id", "disease"), reduction = "pca", dims.use = 1:30, reduction.save = "harmony", assay.use = "SCT")

DefaultAssay(seu) <- "ADT"
VariableFeatures(seu) <- rownames(seu[["ADT"]])
seu <- seu %>% 
    NormalizeData(normalization.method = 'CLR', margin = 2) %>% 
    ScaleData() %>% 
    RunPCA(reduction.name = 'apca', npcs = 20) %>%
    RunHarmony(group.by.vars = c("patient_id", "disease"), reduction = "apca", dims.use = 1:20, reduction.save = "aharmony", assay.use = "ADT")

DefaultAssay(seu) <- "SCT"
```

```{r}
seu <- FindMultiModalNeighbors(
    seu,
    reduction.list = list("harmony", "aharmony"), 
    dims.list = list(1:30, 1:20), 
    modality.weight.name = "RNA.weight"
)

seu <- RunUMAP(seu, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_", seed.use = 42)
seu <- FindClusters(seu, graph.name = "wsnn", algorithm = 1, resolution = 1, verbose = T)
```

```{r}
saveRDS(seu, file = "seu.rds")
```

```{r fig.height=6, fig.width=8}
DimPlot(seu, reduction = "wnn.umap", group.by = c("seurat_clusters"), label = T) + NoLegend()
```
# Identification of cDCs

```{r fig.height=30, fig.width=10}
VlnPlot(seu, group.by = "seurat_clusters", features = c("adt_CD14--M5E2-TSA", "sct_CD14", "adt_CD1c-BDCA1--L161-TSA", "sct_CD1C", "adt_HLA-DR--L243-TSA", "sct_HLA-DRB1", "adt_CD163--GHI-61-TSA", "sct_CD163", "adt_FCER1a--AER-37-TSA", "sct_FCER1A", "sct_FLT3", "sct_CD36", "sct_S100A8", "sct_S100A9", "sct_VCAN", "sct_CLEC4E", "sct_CLEC10A", "sct_FCN1", "sct_C5AR1", "sct_CLEC9A"), pt.size = 0, ncol = 2)
```
```{r fig.height=18, fig.width=10}
FeaturePlot(seu, reduction = "wnn.umap", c("FCER1A", "CD1C", "CLEC10A", "CLEC9A", "CD14", "S100A8",  "CD163", "FLT3", "GZMB", "CD34"), order = T, min.cutoff = "q1", max.cutoff = "q99", ncol = 2)
```
Cluster 5 is cDC2, cluster 28 is cDC1: retain these clusters and re-cluster to identify cDC subpopulations.

# Drill down into cDCs

```{r}
cdc <- subset(seu, subset = seurat_clusters %in% c(5, 28))
```

```{r}
dim(cdc)
```
```{r message=FALSE, warning=FALSE}
plan("sequential")

cdc <- cdc %>% 
    SCTransform(assay = "RNA", new.assay.name = "SCT", do.correct.umi = T, vars.to.regress = c("nCount_RNA", "percent.mt"), variable.features.n = 200, variable.features.rv.th = NULL, return.only.var.genes = F, conserve.memory = F, seed.use = 123) %>% 
    RunPCA(assay = "SCT", npcs = 20) %>% 
    RunHarmony(group.by.vars = c("patient_id", "disease"), reduction = "pca", dims.use = 1:20, reduction.save = "harmony", assay.use = "SCT") %>% 
    RunUMAP(reduction = "harmony", dims = 1:20, seed.use = 42) %>% 
    FindNeighbors(dims = 1:20, reduction = "harmony", force.recalc = TRUE) %>% 
    FindClusters(resolution = 0.3, algorithm = 1)
```

```{r fig.height=4, fig.width=8}
p1 <- DimPlot(cdc, reduction = "umap", group.by = c("seurat_clusters"), label = T) + NoLegend()
p2 <- DimPlot(cdc, reduction = "umap", group.by = c("disease"), label = F)
p1 | p2
rm(list = c("p1", "p2"))
```
```{r fig.height=12, fig.width=12}
FeaturePlot(cdc, reduction = "umap", c("HLA-DRB1","CLEC10A","CLEC9A", "CD1C","FCER1A","CD163","CD14",  "S100A8","CD5"), order = T, min.cutoff = "q5", max.cutoff = "q95", ncol = 3)
```
```{r fig.height=4, fig.width=8}
VlnPlot(cdc, c("HLA-DRB1","CLEC10A","CLEC9A", "CD1C","FCER1A","CD163","CD14","S100A8"), ncol = 4, pt.size = 0) & theme(axis.title.x = element_blank(), axis.text = element_text(size = 10), axis.title = element_text(size = 12))
```
Three cDC subsets are identified:

- cDC1: cluster 4

- DC3: cluster 0

- DC2: clusters 1,2,3

- Clusters 5,6,7? Check if they are doublets

## Doublet identification 

```{r}
sce_cdc <- as.SingleCellExperiment(cdc)
```

### First method: computeDoubletDensity function from scDblFinder package

```{r}
library(scDblFinder)

dbl.dens <- computeDoubletDensity(sce_cdc)
sce_cdc$DoubletScore <- dbl.dens
```

```{r fig.height=3.5, fig.width=5}
plotColData(sce_cdc, x="seurat_clusters", y="DoubletScore", colour_by="seurat_clusters") +
    stat_summary(fun = median, geom='point', size = 10, colour = "black", shape = 95) + 
    NoLegend() +
    theme(axis.text = element_text(size = 10),
          axis.title = element_text(size = 12),
          axis.title.x = element_blank()) 
```
### Second method: cxds function from single cell doublet scoring (scds) package

```{r}
library(scds)

# Annotate doublet using co-expression based doublet scoring:
sce_cdc <- cxds(sce_cdc)
```

```{r fig.width=5, fig.height=3.5}
plotColData(sce_cdc, x="seurat_clusters", y="cxds_score", colour_by="seurat_clusters") + 
    stat_summary(fun = median, geom='point', size = 10, colour = "black", shape = 95)+
    NoLegend() +
    theme(axis.text = element_text(size = 10),
          axis.title = element_text(size = 12),
          axis.title.x = element_blank()) 
```
Both methods assign higher doublet scores to clusters 5, 6 and 7 (in both methods, higher scores indicate more “doublet-like” cells). Hence, remove clusters 5, 6 and 7 and re-cluster again.

# Remove clusters identified as doublets and re-cluster again

```{r message=FALSE, warning=FALSE}
cdc <- subset(cdc, subset = seurat_clusters %in% c(0,1,2,3,4))

plan("sequential")

cdc <- cdc %>% 
    SCTransform(assay = "RNA", new.assay.name = "SCT", do.correct.umi = T, vars.to.regress = c("nCount_RNA", "percent.mt"), variable.features.n = 200, variable.features.rv.th = NULL, return.only.var.genes = F, conserve.memory = F, seed.use = 123) %>% 
    RunPCA(assay = "SCT", npcs = 20) %>% 
    RunHarmony(group.by.vars = c("patient_id", "disease"), reduction = "pca", dims.use = 1:20, reduction.save = "harmony", assay.use = "SCT") %>% 
    RunUMAP(reduction = "harmony", dims = 1:20, seed.use = 42) %>% 
    FindNeighbors(dims = 1:20, reduction = "harmony", force.recalc = TRUE) %>% 
    FindClusters(resolution = 0.5, algorithm = 1)
```

```{r}
dim(cdc)
```
```{r fig.height=3.5, fig.width=6}
p1 <- DimPlot(cdc, reduction = "umap", group.by = c("seurat_clusters"), cols = c('4' = '#ED68ED', '0' = '#00A9FF', '3' = '#59EF16', '2' = '#7CAE00', '1' = '#0CB702', '5' = '#8494FF'), label = T) + NoLegend()

p2 <- DimPlot(cdc, reduction = "umap", group.by = c("disease"), label = F)
p1 | p2
rm(list = c("p1", "p2"))
```
```{r fig.height=10, fig.width=10}
FeaturePlot(cdc, reduction = "umap", c("HLA-DRB1","CLEC10A","CLEC9A","CD1C", "FCER1A", "CD163","CD14","S100A8","CD5"), order = T, min.cutoff = "q5", max.cutoff = "q95", ncol = 3)
```

```{r fig.height=4.5, fig.width=5}
VlnPlot(cdc, c("HLA-DRB1","CLEC10A","CLEC9A","CD14","CD163","S100A8"), pt.size = 0, ncol = 3, 
             cols = c('4' = '#ED68ED', '0' = '#00A9FF', '3' = '#59EF16', '2' = '#7CAE00', '1' = '#0CB702', '5' = '#8494FF'))
```
Three cDC subsets are identified:

- cDC1: cluster 4

- DC2: clusters 1,2,3

- DC3: clusters 0,5

```{r}
cdc$dc <- ifelse(
    cdc$seurat_clusters %in% c(1,2,3), "DC2", ifelse(
    cdc$seurat_clusters %in% c(0,5), "DC3", "cDC1"
))
```

```{r}
saveRDS(cdc, file = "mdc_dataset2.rds")
```

# Pseudobulk differential analysis 

Aggregate DCs by type and donor, then run differential test on aggregated counts. This ensures that the unit of independent replication is the sample (donor), not the cell.

```{r}
sce <- as.SingleCellExperiment(cdc, assay = "SCT")
sce <- sce[rowSums(counts(sce) > 2) >= 5, ]
```

```{r}
t(table(sce$dc, sce$patient_id))
t(table(sce$severity, sce$patient_id, useNA = "always"))
```
## cDC1

### COVID-19 vs Healthy

Remove donors with cDC1 < 10. 

#### DEGs

```{r}
sce_cdc1 <- sce[, (sce$patient_id %in% c("cov03","cov08","cov09","cov11","cov12","cov17","cov18")) & (sce$dc %in% c("cDC1"))]
```

```{r}
sce_cdc1$group_id <- factor(sce_cdc1$disease, levels = c("healthy", "covid"))
sce_cdc1$sample_id <- factor(sce_cdc1$patient_id)

dge_cdc1 <- aggregateAcrossCells(sce_cdc1, id = colData(sce_cdc1)[, c("sample_id")])
dge_cdc1 <- calcNormFactors(dge_cdc1)

mtx_cdc1 <- t(scale(t(edgeR::cpm(dge_cdc1, log = T))))
colnames(mtx_cdc1) <- paste(dge_cdc1$samples$group_id, dge_cdc1$samples$sample_id, sep = "_")

dsn_cdc1 <- model.matrix(~ group_id, dge_cdc1$samples)
dge_cdc1 <- estimateDisp(dge_cdc1, dsn_cdc1)
fit_cdc1 <- glmQLFit(dge_cdc1, dsn_cdc1, robust = T)

res_cdc1 <- glmQLFTest(fit_cdc1, coef = "group_idcovid") %>%
    topTags(n = Inf) %>% 
    `[[`("table") %>%
    `[`(c("logFC", "PValue", "FDR")) %>%
    rownames_to_column("gene") %>%
    as_tibble()
```

```{r}
res_cdc1 %>%
    show_dt()
```

#### GSEA with Hallmark

```{r}
rnk_cdc1 <- setNames(-log10(res_cdc1$PValue) * sign(res_cdc1$logFC), res_cdc1$gene)
gsa_cdc1.hallmark <- fgseaMultilevel(gmt.hallmark, rnk_cdc1, minSize = 10, maxSize = 500, sampleSize = 100)
```

```{r}
gsa_cdc1.hallmark %>%
    as_tibble() %>%
    arrange(pval) %>%
    dplyr::select(pathway, NES, pval, padj, leadingEdge) %>%
    show_dt()
```
#### GSEA with BMT

```{r}
gsa_cdc1.btm <- fgseaMultilevel(gmt.btm, rnk_cdc1, minSize = 10, maxSize = 500, sampleSize = 100)
```

```{r}
gsa_cdc1.btm %>%
    as_tibble() %>%
    arrange(pval) %>%
    dplyr::select(pathway, NES, pval, padj, leadingEdge) %>%
    show_dt()
```

## DC2

###  COVID-19 vs Healthy 

Donors `cov10` removed due to insufficient numbers of cells (< 10).

#### DEGs

```{r}
sce_dc2 <- sce[, (!sce$patient_id %in% c("cov10")) & (sce$dc %in% c("DC2"))]
```

```{r}
sce_dc2$group_id <- factor(sce_dc2$disease, levels = c("healthy", "covid"))
sce_dc2$sample_id <- factor(sce_dc2$patient_id)

dge_dc2 <- aggregateAcrossCells(sce_dc2, id = colData(sce_dc2)[, c("sample_id")])
dge_dc2 <- calcNormFactors(dge_dc2)

mtx_dc2 <- t(scale(t(edgeR::cpm(dge_dc2, log = T))))
colnames(mtx_dc2) <- paste(dge_dc2$samples$group_id, dge_dc2$samples$sample_id, sep = "_")

dsn_dc2 <- model.matrix(~ group_id, dge_dc2$samples)
dge_dc2 <- estimateDisp(dge_dc2, dsn_dc2)
fit_dc2 <- glmQLFit(dge_dc2, dsn_dc2, robust = T)

res_dc2 <- glmQLFTest(fit_dc2, coef = "group_idcovid") %>%
    topTags(n = Inf) %>% 
    `[[`("table") %>%
    `[`(c("logFC", "PValue", "FDR")) %>%
    rownames_to_column("gene") %>%
    as_tibble()
```

```{r}
res_dc2 %>%
    show_dt()
```
#### GSEA with Hallmark

```{r}
rnk_dc2 <- setNames(-log10(res_dc2$PValue) * sign(res_dc2$logFC), res_dc2$gene)
gsa_dc2.hallmark <- fgseaMultilevel(gmt.hallmark, rnk_dc2, minSize = 10, maxSize = 500, sampleSize = 100)
```

```{r}
gsa_dc2.hallmark %>%
    as_tibble() %>%
    arrange(pval) %>%
    dplyr::select(pathway, NES, pval, padj, leadingEdge) %>%
    show_dt()
```

#### GSEA with BTM

```{r}
gsa_dc2.btm <- fgseaMultilevel(gmt.btm, rnk_dc2, minSize = 10, maxSize = 500, sampleSize = 100)
```

```{r}
gsa_dc2.btm %>%
    as_tibble() %>%
    arrange(pval) %>%
    dplyr::select(pathway, NES, pval, padj, leadingEdge) %>%
    show_dt()
```

## DC3

###  COVID-19 vs Healthy 

Donors `cov10` and `cov04` removed due to insufficient numbers of cells (< 10).

#### DEGs

```{r}
sce_dc3 <- sce[, (!sce$patient_id %in% c("cov10","cov04")) & (sce$dc %in% c("DC3"))]
```

```{r}
sce_dc3$group_id <- factor(sce_dc3$disease, levels = c("healthy", "covid"))
sce_dc3$sample_id <- factor(sce_dc3$patient_id)

dge_dc3 <- aggregateAcrossCells(sce_dc3, id = colData(sce_dc3)[, c("sample_id")])
dge_dc3 <- calcNormFactors(dge_dc3)

mtx_dc3 <- t(scale(t(edgeR::cpm(dge_dc3, log = T))))
colnames(mtx_dc3) <- paste(dge_dc3$samples$group_id, dge_dc3$samples$sample_id, sep = "_")

dsn_dc3 <- model.matrix(~ group_id, dge_dc3$samples)
dge_dc3 <- estimateDisp(dge_dc3, dsn_dc3)
fit_dc3 <- glmQLFit(dge_dc3, dsn_dc3, robust = T)

res_dc3 <- glmQLFTest(fit_dc3, coef = "group_idcovid") %>%
    topTags(n = Inf) %>% 
    `[[`("table") %>%
    `[`(c("logFC", "PValue", "FDR")) %>%
    rownames_to_column("gene") %>%
    as_tibble()
```

```{r}
res_dc3 %>%
    show_dt()
```
#### GSEA with Hallmark

```{r}
rnk_dc3 <- setNames(-log10(res_dc3$PValue) * sign(res_dc3$logFC), res_dc3$gene)
gsa_dc3.hallmark <- fgseaMultilevel(gmt.hallmark, rnk_dc3, minSize = 10, maxSize = 500, sampleSize = 100)
```

```{r}
gsa_dc3.hallmark %>%
    as_tibble() %>%
    arrange(pval) %>%
    dplyr::select(pathway, NES, pval, padj, leadingEdge) %>%
    show_dt()
```
#### GSEA with BTM

```{r}
gsa_dc3.btm <- fgseaMultilevel(gmt.btm, rnk_dc3, minSize = 10, maxSize = 500, sampleSize = 100)
```

```{r}
gsa_dc3.btm %>%
    as_tibble() %>%
    arrange(pval) %>%
    dplyr::select(pathway, NES, pval, padj, leadingEdge) %>%
    show_dt()
```
# Heatmap with top 100 DEGs for cDC1, DC2 and DC3

- Rows: Top 100 DEGs for cDC1/DC2/DC3 in COVID19 vs Healthy 

- Columns: samples aggregated by donor and cell type

- Cells: Z-scores, log2 CPMs

Remove RP genes.

## cDC1

```{r}
genes_dc1 <- arrange(res_cdc1, PValue)
genes_dc1 <- genes_dc1[!grepl("^RP[LS]",genes_dc1$gene),]
genes_dc1_top100 <- genes_dc1[1:100, 1]

mtx_dc1.df <- as.data.frame(mtx_cdc1)
mtx_dc1.df$gene <- row.names(mtx_dc1.df)

mtx_dc1.df_top100 <- merge(genes_dc1_top100, mtx_dc1.df, by = 'gene', all.x=TRUE)
row.names(mtx_dc1.df_top100) <- mtx_dc1.df_top100$gene
mtx_dc1.df_top100$gene <- NULL
mtx_dc1.df_top100 <- mtx_dc1.df_top100[,c(2,3,6,7,1,4,5)]
```

```{r fig.height=8, fig.width=3.5}
dend <- dendsort(hclust(dist(mtx_dc1.df_top100), method="ward.D2"))

mtx_dc1.df_top100 %>%
    Heatmap(
        cluster_rows = dend, 
        cluster_columns = F, 
        clustering_method_rows = "ward.D2", 
        row_names_gp = gpar(fontsize = 6),
        col = circlize::colorRamp2(seq(-2.5, 2.5, length.out = 11), rev(brewer.pal(11, "RdYlBu")))
    )
```
## DC2

```{r}
genes_dc2 <- arrange(res_dc2, PValue)
genes_dc2 <- genes_dc2[!grepl("^RP[LS]",genes_dc2$gene),]
genes_dc2_top100 <- genes_dc2[1:100, 1]

mtx_dc2.df <- as.data.frame(mtx_dc2)
mtx_dc2.df$gene <- row.names(mtx_dc2.df)

mtx_dc2.df_top100 <- merge(genes_dc2_top100, mtx_dc2.df, by = 'gene', all.x=TRUE)
row.names(mtx_dc2.df_top100) <- mtx_dc2.df_top100$gene
mtx_dc2.df_top100$gene <- NULL
mtx_dc2.df_top100 <- mtx_dc2.df_top100[,c(5,6,7,10,11,2,3,9,1,4,8)]
```

```{r fig.height=8, fig.width=3.5}
dend <- dendsort(hclust(dist(mtx_dc2.df_top100), method="ward.D2"), isReverse = T)

mtx_dc2.df_top100 %>%
    Heatmap(
        cluster_rows = dend, 
        cluster_columns = F, 
        clustering_method_rows = "ward.D2", 
        row_names_gp = gpar(fontsize = 6),
        col = circlize::colorRamp2(seq(-2.5, 2.5, length.out = 11), rev(brewer.pal(11, "RdYlBu")))
    )
```
## DC3

```{r}
genes_dc3 <- arrange(res_dc3, PValue)
genes_dc3 <- genes_dc3[!grepl("^RP[LS]",genes_dc3$gene),]
genes_dc3_top100 <- genes_dc3[1:100, 1]

mtx_dc3.df <- as.data.frame(mtx_dc3)
mtx_dc3.df$gene <- row.names(mtx_dc3.df)

mtx_dc3.df_top100 <- merge(genes_dc3_top100, mtx_dc3.df, by = 'gene', all.x=TRUE)
row.names(mtx_dc3.df_top100) <- mtx_dc3.df_top100$gene
mtx_dc3.df_top100$gene <- NULL
mtx_dc3.df_top100 <- mtx_dc3.df_top100[,c(4,5,6,9,10,1,2,3,7,8)]
```

```{r fig.height=8, fig.width=3.5}
dend <- dendsort(hclust(dist(mtx_dc3.df_top100), method="ward.D2"))

mtx_dc3.df_top100 %>%
    Heatmap(
        cluster_rows = dend, 
        cluster_columns = F, 
        clustering_method_rows = "ward.D2", 
        row_names_gp = gpar(fontsize = 6),
        col = circlize::colorRamp2(seq(-2.5, 2.5, length.out = 11), rev(brewer.pal(11, "RdYlBu")))
    )
```
# Pathway enrichment visualization

## Hallmark

For each cDC subset, top 15 pathways (ordered by pvalue) are selected, merged and shown.

```{r}
gsa_dc1.hallmark <- arrange(gsa_cdc1.hallmark, pval)
gsa_dc1.hallmark_top <- gsa_dc1.hallmark[1:15, ]

gsa_dc2.hallmark <- arrange(gsa_dc2.hallmark, pval)
gsa_dc2.hallmark_top <- gsa_dc2.hallmark[1:15, ]

gsa_dc3.hallmark <- arrange(gsa_dc3.hallmark, pval)
gsa_dc3.hallmark_top <- gsa_dc3.hallmark[1:15, ] 

pathways.hallmark <- rbind(gsa_dc1.hallmark_top, gsa_dc2.hallmark_top, gsa_dc3.hallmark_top)
pathways.hallmark2 <- pathways.hallmark[!duplicated(pathways.hallmark$pathway),]
pathways.hallmark2 <- pathways.hallmark2[, -c(2:8)]

result_dc1 <- merge(pathways.hallmark2, gsa_dc1.hallmark)
result_dc1$DC_subsets <- 'cDC1'
result_dc1$count <- -log10(result_dc1$pval)

result_dc2 <- merge(pathways.hallmark2, gsa_dc2.hallmark)
result_dc2$DC_subsets <- 'DC2'
result_dc2$count <- -log10(result_dc2$pval) 

result_dc3 <- merge(pathways.hallmark2, gsa_dc3.hallmark)
result_dc3$DC_subsets <- 'DC3'
result_dc3$count <- -log10(result_dc3$pval) 

hallmark <- rbind(result_dc1, result_dc2, result_dc3)
```

```{r}
hallmark$pathway <- gsub('HALLMARK_', '', hallmark$pathway)
hallmark$pathway <- gsub('_', ' ', hallmark$pathway)
```

```{r}
order <- setDT(hallmark)[, lapply(.SD, sum), by = pathway, .SDcols = 'NES']
names(order)[names(order) == "NES"] <- "rowwise_sum_NES"
hallmark <- merge(x = hallmark, y = order, by='pathway', all.x = TRUE)
```

```{r fig.height=5.5, fig.width=5.5}
p <- ggplot(hallmark, aes(x = DC_subsets, y = fct_reorder(pathway,rowwise_sum_NES))) +
    geom_point(aes(size = count, colour = NES)) +
    ylab(NULL) + xlab(NULL) + 
    scale_color_gradientn(colors = c("darkblue", "lightblue", "white", "pink", "darkred"),
                    values = scales::rescale(c(-3.1, -1.2, -0.8, 0.8, 1.2, 3.1)),
                    limits = c(-3.1, 3.1)) +
    labs(size = "-Log10(PValue)") +
    theme(axis.text.x = element_text(angle = 00, hjust = 0.5, size=10), 
          axis.text.y = element_text(size=10),
          legend.title = element_text(size = 9),
          legend.key.size = unit(0.25, "cm"),
          legend.key.width = unit(0.25,"cm"), 
          legend.text=element_text(size=5))

p
```
```{r}
ggsave("hallmark.png", p, width = 5, height = 5, units = c("in"), dpi=600)
```

## BTM

or each cDC subset, top 10 pathways (ordered by pvalue) are selected, merged and shown.

```{r}
gsa_dc1.btm <- arrange(gsa_cdc1.btm, pval)
gsa_dc1.btm_top <- gsa_dc1.btm[1:10, ]

gsa_dc2.btm <- arrange(gsa_dc2.btm, pval)
gsa_dc2.btm_top <- gsa_dc2.btm[1:10, ]

gsa_dc3.btm <- arrange(gsa_dc3.btm, pval)
gsa_dc3.btm_top <- gsa_dc3.btm[1:10, ] 

pathways.btm <- rbind(gsa_dc1.btm_top, gsa_dc2.btm_top, gsa_dc3.btm_top)
pathways.btm2 <- pathways.btm[!duplicated(pathways.btm$pathway),]
pathways.btm2 <- pathways.btm2[, -c(2:8)]

result_dc1 <- merge(pathways.btm2, gsa_dc1.btm)
result_dc1$DC_subsets <- 'cDC1'
result_dc1$count <- -log10(result_dc1$pval)

result_dc2 <- merge(pathways.btm2, gsa_dc2.btm)
result_dc2$DC_subsets <- 'DC2'
result_dc2$count <- -log10(result_dc2$pval) 

result_dc3 <- merge(pathways.btm2, gsa_dc3.btm)
result_dc3$DC_subsets <- 'DC3'
result_dc3$count <- -log10(result_dc3$pval) 

btm <- rbind(result_dc1, result_dc2, result_dc3)
```

```{r}
order <- setDT(btm)[, lapply(.SD, sum), by = pathway, .SDcols = 'NES']
names(order)[names(order) == "NES"] <- "rowwise_sum_NES"
btm <- merge(x = btm, y = order, by='pathway', all.x = TRUE)
```

```{r fig.height=5, fig.width=7}
p <- ggplot(btm, aes(x = DC_subsets, y = fct_reorder(pathway,rowwise_sum_NES))) +
    geom_point(aes(size = count, colour = NES)) +
    ylab(NULL) + xlab(NULL) + 
    scale_color_gradientn(colors = c("darkblue", "lightblue", "white", "pink", "darkred"),
                    values = scales::rescale(c(-2.5, -1.2, -0.8, 0.8, 1.2, 2.5)),
                    limits = c(-2.5, 2.5)) +
    labs(size = "-Log10(PValue)") +
    theme(axis.text.x = element_text(angle = 00, hjust = 0.5, size=10), 
          axis.text.y = element_text(size=10),
          legend.title = element_text(size = 9),
          legend.key.size = unit(0.25, "cm"),
          legend.key.width = unit(0.25,"cm"), 
          legend.text=element_text(size=5))

p
```
```{r}
ggsave("btm.png", p, width = 6.7, height = 4.5, units = c("in"), dpi=600)
```
